service/history/handler.go

```diff
@@ -256,113 +250,78 @@ func (h *Handler) Watch(*healthpb.HealthCheckRequest, healthpb.Health_WatchServe
 
 // RecordActivityTaskHeartbeat - Record Activity Task Heart beat.
 func (h *Handler) RecordActivityTaskHeartbeat(ctx context.Context, request *historyservice.RecordActivityTaskHeartbeatRequest) (_ *historyservice.RecordActivityTaskHeartbeatResponse, retError error) {
-
 	defer log.CapturePanic(h.GetLogger(), &retError)
 	h.startWG.Wait()
 
-	scope := metrics.HistoryRecordActivityTaskHeartbeatScope
-	h.GetMetricsClient().IncCounter(scope, metrics.ServiceRequests)
-	sw := h.GetMetricsClient().StartTimer(scope, metrics.ServiceLatency)
-	defer sw.Stop()
-
 	namespaceID := request.GetNamespaceId()
 	if namespaceID == "" {
-		return nil, h.error(errNamespaceNotSet, scope, namespaceID, "")
-	}
-
-	if ok := h.rateLimiter.Allow(); !ok {
-		return nil, h.error(errHistoryHostThrottle, scope, namespaceID, "")
+		return nil, h.convertError(errNamespaceNotSet)
 	}
 
 	heartbeatRequest := request.HeartbeatRequest
 	taskToken, err0 := h.tokenSerializer.Deserialize(heartbeatRequest.TaskToken)
 	if err0 != nil {
-		return nil, h.error(errDeserializeTaskToken.MessageArgs(err0), scope, namespaceID, "")
+		return nil, h.convertError(errDeserializeTaskToken.MessageArgs(err0))
```

Comment:
`convertError` is another good candidate for interceptor.

Reply:
all common API handler logic goes into interceptor